### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by amirshane**
initial pull request for lens testing files

### CODE REVIEW

File: test_contextual_lenses_learning.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
+# -*- coding: utf-8 -*-
+"""test_contextual_lenses_learning.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1PykmTCyuETlX4UFqoFCRpAJLIAxbAyg2
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:7]: Change this to a one-liner: """Tests for contextual_lenses.py."""

File: test_contextual_lenses_learning.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+import matplotlib.pyplot as plt
+
+import scipy.stats
+
+import sklearn.linear_model
+import sklearn.metrics
+
+import seaborn as sns
+
+"""# Contextual lenses"""
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:41]: move this to the first line in the file and add a little more detail. Also, it should be a complete sentence.

File: test_contextual_lenses_learning.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+
+def max_pool(x, rep_size=None):
+  # Take maximum over sequence length
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:44]: Make these one-line docstrings surrounded in """ quotes and make them full sentences that end with a period.

File: test_contextual_lenses_learning.py [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+  
+  x = nn.relu(x)
+  
+  rep = mean_pool(x, rep_size)
+  
+  return rep
+
+"""# Models"""
+
+def create_optimizer(model, learning_rate, weight_decay):
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:94]: move all of the stuff that's not for contextual lenses to a different file. Could call this train.py or util.py or something.

File: test_contextual_lenses_learning.py [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+  model = nn.Model(module, initial_params)
+  return model
+
+def create_optimizer(model, learning_rate, weight_decay):
+  optimizer_def = optim.Adam(learning_rate=learning_rate, weight_decay=weight_decay)
+  optimizer = optimizer_def.create(model)
+  return optimizer
+
+@jax.jit
+def train_step(optimizer, X, Y):
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:282]: why are there two train_step() functions?

File: test_contextual_lenses_pooling.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+import sklearn.linear_model
+import sklearn.metrics
+
+import seaborn as sns
+
+from absl.testing import parameterized
+
+"""# Contextual lenses"""
+
+def max_pool(x, padding_mask, rep_size=None):
```
Comment (davidBelanger) [inline test_contextual_lenses_pooling.py:45]: you shouldn't be defining max_pool, etc. here. you should be importing from contextual_lenses.py

File: test_contextual_lenses_learning.py [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+
+  preds = jnp.squeeze(optimizer.target(input_data), axis=1)
+
+  train_loss = jnp.mean((preds-output_data)**2)
+
+  return train_loss
+
+"""## No encoding"""
+
+def test_mean():
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:322]: everything above here should be in the library code (util.py/train.py and contextual_lenses.py). Everything below here should be the test. It needs to be wrapped in a test class. I'd using absl.testing.TestCase.

File: test_contextual_lenses_learning.py [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+  
+  optimizer = create_optimizer(model, learning_rate=lr, weight_decay=0.)
+  
+  epochs = 150
+  for epoch in range(epochs):
+    optimizer = train_step(optimizer, input_data, output_data)
+
+  preds = jnp.squeeze(optimizer.target(input_data), axis=1)
+
+  train_loss = jnp.mean((preds-output_data)**2)
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:316]: should have a helper function in utils.py for mse. I'd add some assertions for the input shapes to make sure that you don't hit the bugs you hit last week.

File: test_contextual_lenses_learning.py [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+
+bs = 5
+seq_len = 4
+rep_size = 10
+input_data = jnp.array(np.random.normal(scale=5, size=(bs, seq_len, rep_size)))
+output_data = jnp.array(np.random.normal(size=(bs)))
+
+loss_threshold = 1e-4
+
+def compute_train_loss(model, input_data, output_data, lr=1e-3):
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:306]: this function can be a helper function in the test file.

File: test_contextual_lenses_learning.py [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+
+"""# Tests
+
+## Data
+"""
+
+bs = 5
+seq_len = 4
+rep_size = 10
+input_data = jnp.array(np.random.normal(scale=5, size=(bs, seq_len, rep_size)))
```
Comment (davidBelanger) [inline test_contextual_lenses_learning.py:301]: see below comment about removing all global variables.

**Comment by amirshane** [comment]
Abstracted library with testing scripts for contextual lenses, encoders, and general train utils.

File: test_contextual_lenses/contextual_lenses.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        H = nn.relu(H)
+        G = nn.relu(G)
+      else:
+        H = nn.tanh(H)
+        G = nn.sigmoid(G)
+
+    H, G = H[:, :, 0, :], G[:, :, 0, :]
+    F = H * G + G_0
+
+    F = nn.relu(nn.Dense(F, rep_size))
```
Comment (dmrd) [inline test_contextual_lenses/contextual_lenses.py:113]: Not sure whether this F is necessary. Probably doesn't matter too much either way - curious to experiment once hparam tuning works.

File: test_contextual_lenses/encoders.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
+"""Encoder functions."""
+"""Fixed and learnable transformations for embedding sequences."""
```
Comment (dmrd) [inline test_contextual_lenses/encoders.py:2]: ditto on multiline docstrings

File: test_contextual_lenses/encoders.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    
+    x = jnp.expand_dims(x, axis=2)
+
+    for layer in range(N_layers):
+      features = N_features[layer]
+      kernel_size = (N_kernel_sizes[layer], 1)
+      x = nn.Conv(x, features=features, kernel_size=kernel_size)
+      x = nn.relu(x)
+    
+    x = x[:, :, 0, :]
```
Comment (dmrd) [inline test_contextual_lenses/encoders.py:37]: use jnp.squeeze(x, axis=2) instead to remove dimension (and ensure it is in fact length==1)

File: test_contextual_lenses/encoders.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  """Applies one-hot encoding followed by 1D CNN."""
+
+  one_hots = one_hot_encoder(batch_inds, num_categories, **encoder_fn_kwargs)
+  N_layers, N_features, N_kernel_sizes = list(encoder_fn_kwargs.values())
+  cnn_one_hots = CNN(one_hots, N_layers, N_features, N_kernel_sizes)
+  
+  return cnn_one_hots
+
+
+"""Positional embeddings."""
```
Comment (dmrd) [inline test_contextual_lenses/encoders.py:52]: ditto multiline

File: test_contextual_lenses/test_pooling.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+
+import numpy as np
+
+from absl.testing import parameterized
+from absl.testing import absltest
+
+from contextual_lenses import mean_pool, max_pool
+
+
+"""Test cases."""
```
Comment (dmrd) [inline test_contextual_lenses/test_pooling.py:14]: use # for single line comments.

File: test_contextual_lenses/train_utils.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
+"""Train utils."""
```
Comment (dmrd) [inline test_contextual_lenses/train_utils.py:1]: combine comments

File: test_contextual_lenses/contextual_lenses.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+  if padding_mask is not None:
+    x = x * padding_mask
+    rep = jnp.sum(x, axis=-2) / jnp.sum(padding_mask, axis=-2)
+  else:
+    rep = jnp.mean(x, axis=-2)
+
+  return rep
+
+
+def linear_max_pool(x, padding_mask=None, **reduce_fn_kwargs):
```
Comment (davidBelanger) [inline test_contextual_lenses/contextual_lenses.py:41]: Here and below you should have a named argument rep_size. There's no reason to do the unpacking of generic kwargs on line 45.

File: test_contextual_lenses/contextual_lenses.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+  return rep
+
+
+class GatedConv(nn.Module):
+  """Gated Convolutional lens followed by max pooling,
+     see original paper for details."""
+
+  def apply(self, x, padding_mask=None, **reduce_fn_kwargs):
+    
+    rep_size, M_layers, M_features, M_kernel_sizes, zero_rep_size = list(reduce_fn_kwargs.values())
```
Comment (davidBelanger) [inline test_contextual_lenses/contextual_lenses.py:85]: this unpacking of kwargs is risky bc if the user provided them in a different order, you would get the wrong values. Essentially, you're using *args. You should switch to having all of these be named arguments, such that the function signature is apply(self, x, padding_mask=None, rep_size=10, M_layers=5, ...)
Here, I chose default values for these kwargs. Setting defaults can lead to bugs where you accidentally use default values when you meant to use non-default values. You could have the signature be apply(self, x, padding_mask, rep_size, M_layers, ...)

Comment (davidBelanger) [inline test_contextual_lenses/contextual_lenses.py:85]: fyi these variable names don't follow standard python style for snake case. Generally, variables are lower-case snakecase and classes are camel case. This, for example, should be m_layers.

File: test_contextual_lenses/contextual_lenses.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    H_0 = nn.relu(nn.Dense(x, zero_rep_size))
+    G_0 = nn.relu(nn.Dense(x, zero_rep_size))
+    H, G = jnp.expand_dims(H_0, axis=2), jnp.expand_dims(G_0, axis=2)
+
+    for layer in range(1, M_layers+1):
+      
+      if layer < M_layers:
+        H_features, G_features = M_features[layer-1]
+      else:
+        H_features, G_features = zero_rep_size, zero_rep_size
```
Comment (davidBelanger) [inline test_contextual_lenses/contextual_lenses.py:96]: what are the semantics of zero_rep_size? Is this value going to be zero?

File: test_contextual_lenses/encoders.py [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+    x = x[:, :, 0, :]
+
+    return x
+
+
+def cnn_one_hot_encoder(batch_inds, num_categories=21, **encoder_fn_kwargs):
+  """Applies one-hot encoding followed by 1D CNN."""
+
+  one_hots = one_hot_encoder(batch_inds, num_categories, **encoder_fn_kwargs)
+  N_layers, N_features, N_kernel_sizes = list(encoder_fn_kwargs.values())
```
Comment (davidBelanger) [inline test_contextual_lenses/encoders.py:46]: see above comments about unpacking kwargs

File: test_contextual_lenses/encoders.py [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+
+"""Positional embeddings."""
+"""Code source: https://github.com/google/flax/blob/aff10f032e892e28a1acf4dd4ee9dcc6cd39a606/examples/wmt/models.py."""
+def sinusoidal_init(max_len=2048,
+                    min_scale=1.0,
+                    max_scale=10000.0):
+  """1D Sinusoidal Position Embedding Initializer.
+  Args:
+      max_len: maximum possible length for the input.
+      min_scale: float: minimum frequency-scale in sine grating.
```
Comment (davidBelanger) [inline test_contextual_lenses/encoders.py:60]: what does grating mean?

File: test_contextual_lenses/encoders.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+  
+  max_len = list(encoder_fn_kwargs.values())[0]
+  one_hots = jax.nn.one_hot(batch_inds, num_classes=num_categories)
+  one_hots_pos_emb = AddPositionEmbs(one_hots, max_len=max_len)
+  
+  return one_hots_pos_emb
+
+
+def cnn_one_hot_pos_emb_encoder(batch_inds, num_categories=21, **encoder_fn_kwargs):
+  """Applies one-hot encoding with opsitional embeddings followed by CNN."""
```
Comment (davidBelanger) [inline test_contextual_lenses/encoders.py:151]: spelling

File: test_contextual_lenses/test_learning.py [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+linear_max_pool, linear_mean_pool, gated_conv 
+
+from train_utils import create_optimizer, train_step, \
+create_representation_model
+
+from encoders import one_hot_encoder, cnn_one_hot_encoder, \
+one_hot_pos_emb_encoder, cnn_one_hot_pos_emb_encoder
+
+
+def generate_random_sequences(bs=3, seq_len=12, num_categories=21):
```
Comment (davidBelanger) [inline test_contextual_lenses/test_learning.py:21]: in general, it's better to use longer, more descriptive variable names. I'd replace bs with batch_size

Comment (davidBelanger) [inline test_contextual_lenses/test_learning.py:21]: This function is also generating regression targets, but the function name and docstring don't reflect this. I'd pull the sampling of the random output_data into a different helper function.

File: test_contextual_lenses/test_learning.py [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+  np.random.seed(0)
+  input_data = jnp.array(np.random.randint(0, num_categories-1, size=(bs, seq_len)))
+  
+  np.random.seed(0)
+  output_data = jnp.array(np.random.normal(size=(bs)))
+
+  return input_data, output_data
+
+
+def compute_train_loss(model, input_data, output_data, learning_rate=1e-3, epochs=100):
```
Comment (davidBelanger) [inline test_contextual_lenses/test_learning.py:33]: I'd call this train(). To me, compute_train_loss() would compute the training loss wrt the current model parameters. Here, you're actually optimizing doing training.

File: test_contextual_lenses/test_learning.py [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+  """Fits model to training data and returns train loss."""
+  # print(input_data)
```
Comment (davidBelanger) [inline test_contextual_lenses/test_learning.py:35]: remove

File: test_contextual_lenses/test_learning.py [hunk 26]
```diff
@@ -0,0 +1,454 @@
...
+
+  return train_loss
+
+
+"""Test cases."""
+
+"""One-hot."""
+
+"""Test-case bugged due to optimization issues."""
+# test1 = {
```
Comment (davidBelanger) [inline test_contextual_lenses/test_learning.py:53]: These are fixed, right?

File: test_contextual_lenses/test_pooling.py [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+test1 = {
+          'testcase_name': 'test1',
```
Comment (davidBelanger) [inline test_contextual_lenses/test_pooling.py:16]: If you're using test cases with uninformative names, then you can avoid specifying testcase_name. Just remove that entry from the dict, and use parameterized.parameters() instead of parameterized.named_parameters below.

File: test_contextual_lenses/tape_gfp.py [hunk 28]
```diff
@@ -0,0 +1,673 @@
...
+  return df
+
+
+"""Padding."""
+SEQ_LEN = 237
+def pad_seq(seq, pad_char='-'):
+  
+  padded_seq = seq + pad_char*(SEQ_LEN-len(seq))
+  padding_mask = []
+  for _ in range(len(seq)):
```
Comment (davidBelanger) [inline test_contextual_lenses/tape_gfp.py:50]: you can avoid these for loops. I'd just do padding_mask = np.int32(padded_seq == pad_char).

File: test_contextual_lenses/tape_gfp.py [hunk 29]
```diff
@@ -0,0 +1,673 @@
...
+test_df['padding_masks'] = test_df.primary.apply(lambda x: pad_seq(x)[1])
+test_df.padded_primary.apply(len).describe()
+
+test_df['one_hot_inds'] = test_df.padded_primary.apply(lambda x: residues_to_one_hot_inds(x))
+
+print(test_df.head())
+
+
+"""Data batching."""
+def create_train_data(epochs, train_bs=256):
```
Comment (davidBelanger) [inline test_contextual_lenses/tape_gfp.py:95]: I'd unify the handling of train and test data to make sure it's consistent. I'd have 

create_data_iterator(df, batch_size, shuffle)

and call this for both train and test. The .repeat(epochs) can be done in downstream code.

Comment (davidBelanger) [inline test_contextual_lenses/tape_gfp.py:95]: you should avoid closing over global variables as much as possible. Here, you're closing over the data df. I'd have that be an argument to the function.

File: test_contextual_lenses/tape_gfp.py [hunk 30]
```diff
@@ -0,0 +1,673 @@
...
+  test_batches = tf.data.Dataset.zip((test_one_hot_inds, test_fluorescences))
+  test_batches = test_batches.batch(test_bs).repeat(epochs).as_numpy_iterator()
+
+  return test_batches
+
+test_fluorescences = test_df['log_fluorescence'].values
+
+
+"""Model evaluation."""
+def evaluate(optimizer, test_data, true_fluorescences, plot_title):
```
Comment (davidBelanger) [inline test_contextual_lenses/tape_gfp.py:125]: would be good to have this be able to handle any predictor function, so that you could benchmark against sklearn, for example. All you'd have to change is instead of taking an optimizer arg, take a predict_fn arg, and pass predict_fn=optimizer.target below.

File: test_contextual_lenses/tape_gfp.py [hunk 31]
```diff
@@ -0,0 +1,673 @@
...
+encoder_fn = one_hot_encoder
+encoder_fn_kwargs = {
+    
+}
+reduce_fn = mean_pool
+reduce_fn_kwargs = {
+    
+}
+
+mean_pool_model = create_representation_model(encoder_fn=encoder_fn,
```
Comment (davidBelanger) [inline test_contextual_lenses/tape_gfp.py:204]: Aren't the GFP sequences all the same length? You should definitely benchmark against a reduce_fn that just flattens the one-hots.

For that, it would be good to check the difference in performance from your linear model on top vs. a sklearn linear regression model. They should get the same performance.
